### ISSUE TITLE
[Enhancement] LongPressGestureRecognizer

### ISSUE THREAD

**Issue by PureWeen** [xamarin/Xamarin.Forms#3480]
# LongPressGestureRecognizer

Recognizes when a user presses and holds down one or more fingers for a specified duration of time

_Note: Nothing in this specification is guaranteed to be final; all features, implementations, and interfaces are subject to change._

# API

## Supporting APIs

In order to provide the user with as much available information as possible, we are going to add additional information about all currently active touches. 

#### GestureEventArgs
[Additional Touch Data Attached to Updated events](https://github.com/xamarin/Xamarin.Forms/issues/3479)

## LongPressGestureRecognizer

LongPressGestureRecognizer will provide Long Press Gesture support.

```csharp
public class LongPressGestureRecognizer : GestureRecognizer
{
	public static readonly BindableProperty IsLongPressingProperty;
	public bool IsLongPressing { get; set; }
	
	public static readonly BindableProperty CommandProperty;
	public ICommand Command { get; set; }
	
	public static readonly BindableProperty CommandParameterProperty;
	public object CommandParameter { get; set; }	
	
	// Bindable Property will live on GestureRecognizer
	public ICommand CancelledCommand { get; set; }
	public object CancelledCommandParameter { get; set; }
	
	// Bindable Property will live on GestureRecognizer
	public ICommand StartedCommand { get; set; }
	public object StartedCommandParameter { get; set; }
	
	public static readonly BindableProperty NumberOfTouchesRequiredProperty;
	public int NumberOfTouchesRequired { get; set;}

	public static readonly BindableProperty NumberOfTapsRequiredProperty;
	public int NumberOfTapsRequired { get; set; }

	public static readonly BindableProperty MinimumPressDurationProperty;
	public double MinimumPressDuration { get; set; }

	public static readonly BindableProperty AllowableMovementProperty;
	public uint AllowableMovement { get; set; }

	public event EventHandler LongPressed;
	public event EventHandler<LongPressUpdatedEventArgs> LongPressUpdated;
}
```

### Properties

| API | Description |
| ------------- | ------------- |
| IsLongPressing | One or more fingers have been placed on the screen for the defined duration but the fingers have not released yet. |
| Command | Gets or sets the command to execute once the parameters of the gesture are met, no fingers move outside the allowable range, and then a finger is released.  |
| CommandParameter ||
| CancelledCommand | If the gesture starts then gets cancelled for some reason this command will fire.  |
| CancelledCommandParameter ||
| NumberOfTouchesRequired | Number of touch points required for gesture command to execute (default 1). |
| NumberOfTapsRequired | Number of taps required before gesture is recognized (default 0). If this value is greater than zero the tap count has to be exact for the Gesture to Begin |
| MinimumPressDuration | Minimum amount of time fingers must press down before gesture recognized (default 0.5 seconds). |
| AllowableMovement | Allowable movement for a finger that is down before the gesture doesn't start. |

### Events

| API | Description |
| ------------- | ------------- |
| LongPressed | Occurs when user interaction indicates that the command represented by this gesture should execute. |
| LongPressUpdated | Occurs whenever there are updates from the Long Press. Started - Touches, taps, duration have been met. Running - The touch points are moving around but haven't been lifted. Completed - Finger released. Canceled - Gesture gets canceled before user removes finger |

### LongPressUpdatedEventArgs 

```csharp
public class LongPressUpdatedEventArgs : GestureEventArgs
{	
	public GestureStatus StatusType { get; }
}
```

### Properties

| API | Description |
| ------------- | ------------- |
| GestureStatus | Indicates state of gesture (Started, Running, Canceled, Completed). |

### Examples

```XAML
<Image Source="longpress.jpg">
    <Image.GestureRecognizers>
        <LongPressGestureRecognizer
                IsLongPressing="{Binding IsLongPressing}"
                LongPressed="{Binding OnLongPressed}" />
  </Image.GestureRecognizers>
</Image>
```

A very basic default setup for a Long Press Gesture. This will begin if the user places a single finger on the image and doesn't remove it. Once the finger is removed the OnLongPressed will fire.

***

```XAML
<Image Source="TapTapHoldRelease.jpg">
    <Image.GestureRecognizers>
        <LongPressGestureRecognizer
                IsLongPressing="{Binding IsLongPressing}"
		NumberOfTapsRequired="3"
                Command="{Binding OnLongPressed}" />
  </Image.GestureRecognizers>
</Image>
```

In this example the LongPressGesture fires if the user taps the image twice before leaving the finger on the image. A tap is defined by touching the image and then quickly lifting the finger off the image. 

Down => Up => Down => Up => Down and Hold (IsLongPressing fires) => Move Finger around (Updated event fires with Status Running) => Up (LongPressed Fires and Gesture Completes)

If the user taps the image 3 times and then doesn't lift the Gesture Recognizer will not Begin.

***

```XAML
<Image Source="ThreeFingers.jpg">
    <Image.GestureRecognizers>
        <LongPressGestureRecognizer
                IsLongPressing="{Binding IsLongPressing}"
		NumberOfTouchesRequired="3"
                Command="{Binding OnLongPressed}" />
  </Image.GestureRecognizers>
</Image>
```
Here three fingers have to simultaneously touch the view and remain on the view for > 0.5 seconds before the gesture begins

**Comment by AwsomeCode** [comment]
+1

**Comment by PureWeen** [comment]
The spec follows pretty closely with how UILongPressGestures work in iOS. Not to say we can't make it our own or better :-) 

https://developer.apple.com/documentation/uikit/uilongpressgesturerecognizer

The idea of the command not firing is that Long press is a continuous gesture. Just long pressing something in a menu is one particular use case but the other use case is wanting to long press and then move something around. 

> I guess you could achieve this by listening on PropChanged of IsLongPressing, but that would be really awkward.

I don't feel like it's that awkward.  If you don't care about the continuous nature of the gesture then you just fire whatever you need from your VM when IsLongPressing goes from false to true or inside the StartedCommand. 

Some additional thoughts

- Rename Command to FinishedCommand would align better with what's happening?
- Add a property *IsContinuous* or a *Mode* this way you can articulate to the gesture that you don't care about continuous motion and it'll just "cancel" or "finish" the gesture right after it starts

And it is not being worked on right now. If you want to start on it that would be awesome!!

**Comment by bruzkovsky** [comment]
Thanks for pointing me to that iOS spec! I did not think of a "long press and drag" scenario, but it makes sense now :smile:

And yes, the StartedCommand is what I was missing. So it is executing when these four conditions are met:
1. NumberOfTouchesRequired is matched
2. NumberOfTapsRequired is matched
3. MinimumPressDuration is matched
4. AllowableMovement is not exceeded

If that is true, then it would not be awkward to use the StartedCommand for e.g. showing a context menu.

>Rename Command to FinishedCommand would align better with what's happening?

Renaming to FinishedCommand would totally make sense. Now that I think of it, this could be a pattern for all continuous gestures (Pan, Pinch, Multi-Tap). I noticed StartedCommand and CanceledCommand in your other GR-Specs, so maybe you could add the FinishedCommand there, too.

If the spec is finished, I can start to work on this end of next week.

**Comment by PureWeen** [comment]
> And yes, the StartedCommand is what I was missing. So it is executing when these four conditions are met:

Yes. Those things indicate that the gesture needs to enter a "Started" phase so now it's tied to those touch points and being tracked

> If that is true, then it would not be awkward to use the StartedCommand for e.g. showing a context menu.

Why?  The only "nice to have" I can think of would be a way to cancel the gesture after the start happens but I don't think that's super necessary.  The start is only going to fire once and the continuation of the gesture shouldn't really effect anything.  

> Renaming to FinishedCommand would totally make sense. Now that I think of it, this could be a pattern for all continuous gestures (Pan, Pinch, Multi-Tap). I noticed StartedCommand and CanceledCommand in your other GR-Specs, so maybe you could add the FinishedCommand there, too.

That's a good thought... But let's see are we locked in anywhere right now with using Command. I want to make sure we're consistent :-)  Tap and Swipe look to be the only existing gestures with commands and they aren't continuous. Pinch looks like it needs a spec that's similar to Pan.  

That seems like a good change. It makes more sense to be called Finished when the gesture is continuous. *Command* is kind of meaningless with something that's continuous and doesn't have a defined start/end point.

I'll update the specs....  

> If the spec is finished, I can start to work on this end of next week.

That would be awesome!! If you start working on this just let me know so there isn't duplicate work going on!

**Comment by PureWeen** [comment]
Reference implementation
https://github.com/xamarin/Xamarin.Forms/issues/6338

**Comment by yanshouwang** [comment]
Please tell me how long it will takes for this feature to release? It's really useful and necessary

**Comment by MappingSteve** [comment]
@PureWeen - FWIW, the easiest current workaround I know of is Alex Dunn's https://alexdunn.org/2017/12/27/xamarin-tip-xamarin-forms-long-press-effect/

**Comment by FabienBehra38** [comment]
This feature will be available soon ? Because Alex Dunn's soluce isn't on UWP.

**Comment by KSemenenko** [comment]
Hello everyone, I really need this functional.
I would also like a press and release event.
Maybe I can do this task? @PureWeen  @samhouts

**Comment by samhouts** [comment]
@KSemenenko We'd be thrilled! Thanks!

**Comment by Zure1** [comment]
@samhouts Does removing from "**In Progress** in **vNext (4.4.0)**" mean that this enhancement will be available in XF 4.4? Because I can't find it in **In Review** or **Closed**.

**Comment by jfversluis** [comment]
I think the important line is the one above

> added this to In Progress in vNext+1 (master)

So this has moved up to 4.5 at least

**Comment by ske66** [comment]
Just wanted to check where this is at. I think it was originally meant for 4.5 but now has it been pushed back to 4.6? The naming of the projects is a little bit confusing

**Comment by jfversluis** [comment]
Since 4.5 stable is out no new APIs will be added. This will go into 4.6 at the earliest. What do you find confusing?

**Comment by ske66** [comment]
> Since 4.5 stable is out no new APIs will be added. This will go into 4.6 at the earliest. What do you find confusing?

Originally vNext(4.6) said vNext(+1) which either sounds like 4.6 or 4.7. That's what was confusing about the naming

**Comment by abdul-wasey** [comment]
That is very useful and commonly used feature in Mobile App development. 
I don't know why this issue is not closed even in 2 years of time period.
Hope that I will see this feature in MAUI ;)

**Comment by chrisgull** [comment]
Sorely needed.

**Comment by jfversluis** [comment]
@chrisgull and everyone, I know, you probably think this should be in Forms/.NET MAUI. But to unblock you right now, we have this in the [Xamarin Community Toolkit](https://aka.ms/xct) that might be worth checking out

**Comment by MichaelRumpler** [comment]
This is also included in [MR.Gestures](https://www.mrgestures.com/) since 2015.

**Comment by aschuhardt** [comment]
Did this ever make it into MAUI?

**Comment by NicolasBrondin** [comment]
@aschuhardt Apparently not.
